sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/InstrumentationScope.java

```diff
@@ -0,0 +1,111 @@
+
+package com.azure.messaging.eventhubs.implementation.instrumentation;
+
+import com.azure.core.util.Context;
+
+import java.time.Instant;
+import java.util.function.BiConsumer;
+
+import static com.azure.messaging.eventhubs.implementation.instrumentation.InstrumentationUtils.CANCELLED_ERROR_TYPE_VALUE;
+
+/**
+ * Helper class that holds state necessary for the operation instrumentation.
+ */
+public final class InstrumentationScope implements AutoCloseable {
+    private final EventHubsMetricsProvider meter;
+    private final EventHubsTracer tracer;
+    private final boolean isEnabled;
+    private final BiConsumer<EventHubsMetricsProvider, InstrumentationScope> reportMetricsCallback;
+    private Instant startTime;
+    private Throwable error;
+    private String errorType;
+    private Context span = Context.NONE;
+    private AutoCloseable spanScope;
+    private boolean closed = false;
+
+    public InstrumentationScope(EventHubsTracer tracer,
+                                EventHubsMetricsProvider meter,
+                                BiConsumer<EventHubsMetricsProvider, InstrumentationScope> reportMetricsCallback) {
+        this.tracer = tracer;
+        this.meter = meter;
+        this.isEnabled = (tracer != null && tracer.isEnabled()) || (meter != null && meter.isEnabled());
+        if (meter != null && meter.isEnabled()) { // micro-optimization
+            this.startTime = Instant.now();
+        }
+        this.reportMetricsCallback = reportMetricsCallback;
+    }
+
+    public boolean isEnabled() {
+        return isEnabled;
+    }
+
+    public InstrumentationScope setError(Throwable error) {
+        if (isEnabled) {
+            this.error = error;
+        }
+        return this;
+    }
+
+    public InstrumentationScope setSpan(Context span) {
+        if (isEnabled) {
+            this.span = span;
+        }
+        return this;
+    }
+
+    public InstrumentationScope setCancelled() {
+        // Complicated calls can result in error followed by cancellation. We shouldn't track them twice.
+        // don't trust me? try this:
+        // Flux.fromIterable(Collections.singletonList("event"))
+        //     .flatMap(event -> Mono.error(new RuntimeException("boom"))
+        //         .doOnError(e -> System.out.println("Error"))
+        //         .doOnCancel(() -> System.out.println("Cancel")))
+        //     .blockLast();
+        if (isEnabled && error == null) {
+            errorType = CANCELLED_ERROR_TYPE_VALUE;
+        }
+
+        return this;
+    }
+
+    public Instant getStartTime() {
+        return startTime;
+    }
+
+    public Context getSpan() {
+        return span;
+    }
+
+    public Throwable getError() {
+        return error;
+    }
+
+    public String getErrorType() {
+        if (errorType == null) {
+            errorType = InstrumentationUtils.getErrorType(error);
+        }
+        return errorType;
+    }
+
+    @Override
+    public void close() {
+        if (!closed) {
```

Comment:
since we've this guard check, do we know if there a potential for `close()` to be called from two different threads? if so, wondering `closed` need to be `volatile` or we're safe

Reply:
it should not be executed concurrently, but let's stay on the safe side and assume it could be. So changed to volatile. Thank you!